### DISCUSSION TITLE
No responses with await response.json() when using Jest & undici polyfill

### DISCUSSION THREAD

**Discussion by jeffwong14** [mswjs/msw#1829] [Q&A] [answered]
The same code worked in version 1.3.2 but not working after changed to 2
I followed the migration guide to change the rest.get -> http.get , return HttpResponse.json(xxx)
But the test cases still don't have response.

So I simulate the case out with simple code 

The fetch response is able to print out (added below). However, my code stopped around await resp.json()

Could anyone told me what's wrong with my case ?

[Code Sandbox](https://codesandbox.io/p/sandbox/gifted-liskov-z5qwrn?file=%2Fpackage.json%3A6%2C19&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clohxk2dx00073b6ifydiey2m%2522%252C%2522sizes%2522%253A%255B70.27548209366391%252C29.72451790633609%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clohxk2dx00033b6i7vmf3df8%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clohxk2dx00053b6i4u7of2mu%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clohxk2dx00063b6iol903yi2%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B100%252C0%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clohxk2dx00033b6i7vmf3df8%2522%253A%257B%2522id%2522%253A%2522clohxk2dx00033b6i7vmf3df8%2522%252C%2522activeTabId%2522%253A%2522clohzocjn016b3b6iiex808hd%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FgetData.test.ts%2522%252C%2522id%2522%253A%2522clohxrt8s00og3b6i4ke9g3aq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fjest.config.js%2522%252C%2522id%2522%253A%2522clohziz9i00xu3b6ig0b7sk3z%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fjest.polyfills.js%2522%252C%2522id%2522%253A%2522clohzocjn016b3b6iiex808hd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fjest.setup.js%2522%252C%2522id%2522%253A%2522clohzrho801v23b6iqz8h6trg%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522id%2522%253A%2522clohzzq2403733b6i1a3fm5lw%2522%252C%2522mode%2522%253A%2522temporary%2522%257D%255D%257D%252C%2522clohxk2dx00063b6iol903yi2%2522%253A%257B%2522id%2522%253A%2522clohxk2dx00063b6iol903yi2%2522%252C%2522activeTabId%2522%253A%2522clohxm93800cy3b6imhxhv61r%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522SANDBOX_INFO%2522%252C%2522id%2522%253A%2522clohxm93800cy3b6imhxhv61r%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522SYSTEM_METRICS%2522%252C%2522id%2522%253A%2522clohzpo1h01jj3b6it3oq3fiz%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522clohxk2dx00053b6i4u7of2mu%2522%253A%257B%2522id%2522%253A%2522clohxk2dx00053b6i4u7of2mu%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clohxkvk6006g3b6i3pfeync4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clohxkvna00dcefg66i1rgyc0%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clohxkvk6006g3b6i3pfeync4%2522%257D%257D%252C%2522showDevtools%2522%253Afalse%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A26.27596439169139%257D)

My case 
```
import { HttpResponse, http } from "msw";
import { setupServer } from "msw/node";

const hardCodeNow = new Date(2023, 8, 1, 0, 0, 0, 0);

const thisHost = "http://localhost";

const generateResp = () =>
  http.get(`${thisHost}/picture`, () => {
    return HttpResponse.json({ result: "success" }, { status: 200 });
  });

const server = setupServer(generateResp());

beforeAll(() => {
  server.listen();
  server.events.on("request:start", ({ request }) => {
    console.log("MSW intercepted:", request.method, request.url);
  });
});

beforeEach(() => {
  jest.useFakeTimers({
    now: hardCodeNow,
  });
  jest.resetModules();
  sessionStorage.clear();
  localStorage.clear();
});

afterEach(() => {
  jest.clearAllMocks();
  jest.runOnlyPendingTimers();
  jest.useRealTimers();
  server.resetHandlers();
});

afterAll(() => {
  server.close();
});

describe("Test MSW", () => {
  it("MSW 1", async () => {
    server.use(generateResp());
    const resp = await fetch(`${thisHost}/picture`);
    if (!resp.ok) {
      throw new Error("Response not ok");
    }
    console.log(resp);
    const x = await resp.json();
// ^^^ code stopped here 
    console.log(x);
  });
});
```

The libraries I used
```
  "devDependencies": {
    "@testing-library/jest-dom": "6.1.4",
    "@testing-library/react": "14.0.0",
    "jest": "29.7.0",
    "jest-environment-jsdom": "29.7.0",
    "jest-localstorage-mock": "2.4.26",
    "msw": "2.0.2",
    "typescript": "5.2.2",
    "undici": "^5.27.0"
  },
  "dependencies": {
    "next": "^14.0.1",
    "react": "18.2.0"
  }
```
jest.config
``` 
const nextJest = require("next/jest");

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: "./",
});

// Add any custom config to be passed to Jest
const customJestConfig = {
  resetMocks: false,
  workerIdleMemoryLimit: "512MB",
  setupFilesAfterEnv: ["<rootDir>/jest.setup.js"],
  setupFiles: ["<rootDir>/jest.polyfills.js"],
  testEnvironmentOptions: {
    customExportConditions: [""],
  },
  transform: {},
  testEnvironment: "jest-environment-jsdom",
};

// createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
module.exports = createJestConfig(customJestConfig);
```
Polyfill
```
// jest.polyfills.js
const { TextEncoder, TextDecoder } = require("node:util");

Reflect.set(globalThis, "TextEncoder", TextEncoder);
Reflect.set(globalThis, "TextDecoder", TextDecoder);

const { Blob } = require("node:buffer");
const { fetch, Request, Response, Headers, FormData } = require("undici");

Reflect.set(globalThis, "fetch", fetch);
Reflect.set(globalThis, "Blob", Blob);
Reflect.set(globalThis, "Request", Request);
Reflect.set(globalThis, "Response", Response);
Reflect.set(globalThis, "Headers", Headers);
Reflect.set(globalThis, "FormData", FormData);
```

Response Result
```
Response {
      url: 'http://localhost/picture',
      [Symbol(realm)]: { settingsObject: {} },
      [Symbol(state)]: {
        aborted: false,
        rangeRequested: false,
        timingAllowPassed: false,
        requestIncludesCredentials: false,
        type: 'default',
        status: 200,
        timingInfo: null,
        cacheState: '',
        statusText: 'OK',
        headersList: HeadersList {
          cookies: null,
          [Symbol(headers map)]: [Map],
          [Symbol(headers map sorted)]: null
        },
        urlList: [],
        body: { stream: undefined, source: null, length: null }
      },
      [Symbol(headers)]: HeadersList {
        cookies: null,
        [Symbol(headers map)]: Map(1) { 'content-type' => [Object] },
        [Symbol(headers map sorted)]: null
      }
    }
```

**Comment by jeffwong14** [comment]
I found I should be in the same situation as #1830
As the problem resolved if i removed all the fakeTimers related codes

**Comment by mattcosta7** [comment, marked as answer]
@jeffwong14  - I believe the issue stems from an implementation of fake timers in jest combined with the way node's fetch implementation, unidici works. 

this doesn't seem to be something we can fix directly in MSW as far as I can tell, but instead we would need changes in undici.

I have more discovery as well as 2 workarounds here: https://github.com/mswjs/msw/issues/1830#issuecomment-1793470397

--- 

And the relevant workarounds copied here:

1. use legacy fake times

```ts
  jest.useFakeTimers({
    now: hardCodeNow,
    legacyFakeTimers: true,
  });
```

2. avoid faking queueMicrotask

```ts
  jest.useFakeTimers({
    now: hardCodeNow,
    doNotFake: ["queueMicrotask"],
  });
```

**Reply by mattcosta7** [reply]
also, thank you for the sandbox, which really helped me while digging into this!

**Reply by jeffwong14** [reply]
Many thanks @mattcosta7 for your advice
Really appreciated :)
